File: scase-pulsar/src/main/scala/io/jobial/scase/pulsar/PulsarContext.scala
package io.jobial.scase.pulsar

import cats.effect.IO
import cats.implicits.catsSyntaxFlatMapOps
import org.apache.pulsar.client.api.PulsarClient
import scala.concurrent.ExecutionContext.global

case class PulsarContext(
  host: String = "localhost",
  port: Int = 6650,
  tenant: String = "public",
  namespace: String = "default",
  useDaemonThreadsInClient: Boolean = true
) {
  val brokerUrl = s"pulsar://$host:$port"

  def createClient =
    PulsarClient
      .builder
      .serviceUrl(brokerUrl)
      .build

  lazy val client =
    if (useDaemonThreadsInClient) {
      // Hack to initialize the Pulsar client on a daemon thread so that it creates daemon threads itself...
      implicit val contextShift = IO.contextShift(global)
      (IO(createClient).start >>= (_.join)).unsafeRunSync()
    } else createClient

  val persistentScheme = "persistent:"

  val nonPersistentScheme = "non-persistent:"

  def fullyQualifiedTopicName(topic: String) =
    if (topic.startsWith(persistentScheme) || topic.startsWith(nonPersistentScheme))
      topic
    else
      s"$persistentScheme//$tenant/$namespace/$topic"

}

File: scase-pulsar/src/main/java/io/jobial/scase/pulsar/javadsl/PulsarContext.java
package io.jobial.scase.pulsar.javadsl;

public class PulsarContext {

    private volatile io.jobial.scase.pulsar.PulsarContext context = io.jobial.scase.pulsar.javadsl.package$.MODULE$.defaultPulsarContext();

    public PulsarContext() {
    }

    public PulsarContext(io.jobial.scase.pulsar.PulsarContext context) {
        this.context = context;
    }

    public io.jobial.scase.pulsar.PulsarContext getContext() {
        return context;
    }
    
    public PulsarContext withHost(String host) {
        return new PulsarContext(context.copy(host, context.port(), context.tenant(), context.namespace(), context.useDaemonThreadsInClient()));
    }

    public PulsarContext withPort(int port) {
        return new PulsarContext(context.copy(context.host(), port, context.tenant(), context.namespace(), context.useDaemonThreadsInClient()));
    }

    public PulsarContext withTenant(String tenant) {
        return new PulsarContext(context.copy(context.host(), context.port(), tenant, context.namespace(), context.useDaemonThreadsInClient()));
    }

    public PulsarContext withNamespace(String namespace) {
        return new PulsarContext(context.copy(context.host(), context.port(), context.tenant(), namespace, context.useDaemonThreadsInClient()));
    }

    public PulsarContext withUseDaemonThreadsInClient(Boolean useDaemonThreadsInClient) {
        return new PulsarContext(context.copy(context.host(), context.port(), context.tenant(), context.namespace(), useDaemonThreadsInClient));
    }
}
